File: services/worker-py/src/jobforge_worker/handlers/webhook_deliver.py
"""Webhook Delivery Connector - Python implementation."""

import hashlib
import hmac
import json
import os
import time
from datetime import UTC, datetime
from ipaddress import ip_address, ip_network
from typing import Any
from urllib.parse import urlparse

import httpx
from pydantic import BaseModel, Field, field_validator


class WebhookDeliverPayload(BaseModel):
    """Webhook delivery job payload."""

    target_url: str
    event_type: str = Field(min_length=1)
    event_id: str  # UUID string
    data: dict[str, Any]
    secret_ref: str | None = None
    signature_algo: str = "sha256"
    timeout_ms: int = Field(default=10_000, ge=1, le=60_000)

    @field_validator("signature_algo")
    @classmethod
    def validate_algo(cls, v: str) -> str:
        """Validate signature algorithm."""
        if v not in ["sha256", "sha512"]:
            raise ValueError("signature_algo must be sha256 or sha512")
        return v


class WebhookDeliverResult(BaseModel):
    """Webhook delivery result."""

    delivered: bool
    status: int
    duration_ms: int
    response_preview: str
    signature: str | None
    timestamp: str


PRIVATE_NETWORKS = [
    ip_network("10.0.0.0/8"),
    ip_network("127.0.0.0/8"),
    ip_network("169.254.0.0/16"),
    ip_network("172.16.0.0/12"),
    ip_network("192.168.0.0/16"),
    ip_network("::1/128"),
    ip_network("fc00::/7"),
    ip_network("fe80::/10"),
]


def assert_safe_webhook_target(target_url: str) -> None:
    parsed = urlparse(target_url)
    if parsed.scheme not in {"http", "https"}:
        raise ValueError(f"Unsupported webhook protocol: {parsed.scheme}")

    hostname = (parsed.hostname or "").lower()
    if not hostname:
        raise ValueError("Webhook target hostname is required")

    if hostname in {"localhost"} or hostname.endswith((".localhost", ".local", ".internal")):
        raise ValueError(f"Unsafe webhook target hostname: {hostname}")

    try:
        ip = ip_address(hostname)
    except ValueError:
        return

    if any(ip in network for network in PRIVATE_NETWORKS):
        raise ValueError(f"Unsafe webhook target IP: {hostname}")


def generate_signature(payload: str, secret: str, algo: str) -> str:
    """
    Generate HMAC signature for webhook payload.

    Args:
        payload: JSON payload string
        secret: Signing secret
        algo: Hash algorithm (sha256 or sha512)

    Returns:
        Hex-encoded signature
    """
    hash_func = hashlib.sha256 if algo == "sha256" else hashlib.sha512
    signature = hmac.new(secret.encode(), payload.encode(), hash_func)
    return signature.hexdigest()


def webhook_deliver_handler(payload: dict[str, Any], context: dict[str, Any]) -> dict[str, Any]:
    """
    Webhook delivery handler.

    Args:
        payload: Job payload
        context: Job context

    Returns:
        Delivery result
    """
    validated = WebhookDeliverPayload.model_validate(payload)
    assert_safe_webhook_target(validated.target_url)

    timestamp = datetime.now(UTC).isoformat()

    # Prepare webhook payload
    webhook_payload = {
        "event_type": validated.event_type,
        "event_id": validated.event_id,
        "timestamp": timestamp,
        "data": validated.data,
    }

    payload_string = json.dumps(webhook_payload, separators=(",", ":"))

    # Generate signature if secret provided
    signature: str | None = None
    headers = {
        "Content-Type": "application/json",
        "User-Agent": "JobForge-Webhook/1.0",
        "X-JobForge-Event": validated.event_type,
        "X-JobForge-Event-ID": validated.event_id,
        "X-JobForge-Timestamp": timestamp,
        "X-JobForge-Delivery-Attempt": str(context.get("attempt_no", 1)),
    }

    if validated.secret_ref:
        # Fetch secret from environment using secret_ref as key
        secret = os.environ.get(validated.secret_ref)
        if not secret:
            raise ValueError(f"Secret not found: {validated.secret_ref}")

        signature = generate_signature(payload_string, secret, validated.signature_algo)
        headers["X-JobForge-Signature"] = f"{validated.signature_algo}={signature}"

    start_time = time.time()

    try:
        with httpx.Client(timeout=validated.timeout_ms / 1000.0) as client:
            response = client.post(
                validated.target_url,
                headers=headers,
                content=payload_string,
            )

        duration_ms = int((time.time() - start_time) * 1000)

        response_text = response.text
        response_preview = (
            response_text[:500] + "... (truncated)"
            if len(response_text) > 500
            else response_text
        )

        result = WebhookDeliverResult(
            delivered=response.is_success,
            status=response.status_code,
            duration_ms=duration_ms,
            response_preview=response_preview,
            signature=signature,
            timestamp=timestamp,
        )

    except Exception as e:
        duration_ms = int((time.time() - start_time) * 1000)

        result = WebhookDeliverResult(
            delivered=False,
            status=0,
            duration_ms=duration_ms,
            response_preview=str(e),
            signature=signature,
            timestamp=timestamp,
        )

    return result.model_dump()

File: services/worker-ts/src/handlers/webhook-deliver.ts
/**
 * Webhook Delivery Connector
 * Delivers webhooks with HMAC signing and retry tracking
 */

import type { JobContext } from '@jobforge/shared'
import { z } from 'zod'
import { createHmac } from 'crypto'
import { isIP } from 'net'
import { readBodyPreview } from './response-preview'

const WebhookDeliverPayloadSchema = z.object({
  target_url: z.string().url(),
  event_type: z.string().min(1),
  event_id: z.string().uuid(),
  data: z.record(z.unknown()),
  secret_ref: z.string().optional(),
  signature_algo: z.enum(['sha256', 'sha512']).default('sha256'),
  timeout_ms: z.number().int().positive().max(60_000).default(10_000),
})

export type WebhookDeliverPayload = z.infer<typeof WebhookDeliverPayloadSchema>

export interface WebhookDeliverResult {
  delivered: boolean
  status: number
  duration_ms: number
  response_preview: string
  signature: string | null
  timestamp: string
}

const PRIVATE_HOSTNAMES = new Set(['localhost'])

function isPrivateIpv4(hostname: string): boolean {
  const [a, b] = hostname.split('.').map((segment) => Number(segment))
  if ([a, b].some((value) => Number.isNaN(value))) {
    return false
  }
  if (a === 10 || a === 127 || a === 0) {
    return true
  }
  if (a === 169 && b === 254) {
    return true
  }
  if (a === 192 && b === 168) {
    return true
  }
  if (a === 172 && b >= 16 && b <= 31) {
    return true
  }
  return false
}

function isPrivateIpv6(hostname: string): boolean {
  const normalized = hostname.toLowerCase()
  return (
    normalized === '::1' ||
    normalized.startsWith('fc') ||
    normalized.startsWith('fd') ||
    normalized.startsWith('fe80')
  )
}

function assertSafeWebhookTarget(targetUrl: string): void {
  const url = new URL(targetUrl)
  if (url.protocol !== 'https:' && url.protocol !== 'http:') {
    throw new Error(`Unsupported webhook protocol: ${url.protocol}`)
  }

  const hostname = url.hostname.toLowerCase()
  if (
    PRIVATE_HOSTNAMES.has(hostname) ||
    hostname.endsWith('.localhost') ||
    hostname.endsWith('.local') ||
    hostname.endsWith('.internal')
  ) {
    throw new Error(`Unsafe webhook target hostname: ${hostname}`)
  }

  const ipType = isIP(hostname)
  if (ipType === 4 && isPrivateIpv4(hostname)) {
    throw new Error(`Unsafe webhook target IP: ${hostname}`)
  }
  if (ipType === 6 && isPrivateIpv6(hostname)) {
    throw new Error(`Unsafe webhook target IP: ${hostname}`)
  }
}

/**
 * Generate HMAC signature for webhook payload
 */
function generateSignature(payload: string, secret: string, algo: 'sha256' | 'sha512'): string {
  const hmac = createHmac(algo, secret)
  hmac.update(payload)
  return hmac.digest('hex')
}

/**
 * Webhook Delivery Handler
 */
export async function webhookDeliverHandler(
  payload: unknown,
  context: JobContext
): Promise<WebhookDeliverResult> {
  const validated = WebhookDeliverPayloadSchema.parse(payload)
  assertSafeWebhookTarget(validated.target_url)

  const timestamp = new Date().toISOString()

  // Prepare webhook payload
  const webhookPayload = {
    event_type: validated.event_type,
    event_id: validated.event_id,
    timestamp,
    data: validated.data,
  }

  const payloadString = JSON.stringify(webhookPayload)

  // Generate signature if secret provided
  let signature: string | null = null
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'User-Agent': 'JobForge-Webhook/1.0',
    'X-JobForge-Event': validated.event_type,
    'X-JobForge-Event-ID': validated.event_id,
    'X-JobForge-Timestamp': timestamp,
    'X-JobForge-Delivery-Attempt': String(context.attempt_no),
  }

  if (validated.secret_ref) {
    // In production, fetch secret from secure store using secret_ref
    // For now, read from env var with secret_ref as key
    const secret = process.env[validated.secret_ref]
    if (!secret) {
      throw new Error(`Secret not found: ${validated.secret_ref}`)
    }

    signature = generateSignature(payloadString, secret, validated.signature_algo)
    headers['X-JobForge-Signature'] = `${validated.signature_algo}=${signature}`
  }

  const startTime = Date.now()

  try {
    const response = await fetch(validated.target_url, {
      method: 'POST',
      headers,
      body: payloadString,
      signal: AbortSignal.timeout(validated.timeout_ms),
    })

    const duration_ms = Date.now() - startTime

    const responsePreview = await readBodyPreview(response, 500)

    return {
      delivered: response.ok,
      status: response.status,
      duration_ms,
      response_preview: responsePreview.bodyPreview,
      signature,
      timestamp,
    }
  } catch (error) {
    const duration_ms = Date.now() - startTime

    // Return partial result for retryable errors
    return {
      delivered: false,
      status: 0,
      duration_ms,
      response_preview: error instanceof Error ? error.message : String(error),
      signature,
      timestamp,
    }
  }
}
